### PR TITLE
Add Vapid support and assorted fixes

### PR DESCRIPTION
**Pull request by costinm**
The 'integration' tests are passing, but likely to fail when the subscriptions expire - 
a more 'end-to-end' solution that includes registration is needed.

### CODE REVIEW

**Comment by wibblymat** [comment]
Hi @costinm! Sorry for wait due to the long weekend. Reviewing now.

File: webpush/encrypt.go [hunk 1]
```diff
@@ -139,36 +142,64 @@ type EncryptionResult struct {
 	ServerPublicKey []byte
 }
 
+// Decrypt an encrypted messages.
```
Comment (wibblymat) [inline webpush/encrypt.go]: What's the use case for adding decryption? AFAICT only the UA has any legitimate reason to decrypt the message.

File: webpush/encrypt.go [hunk 2]
```diff
@@ -139,36 +142,64 @@ type EncryptionResult struct {
...
+	nonce := newNonce(ctx, crypt.Salt, prk)
+
+	plain, err = decrypt(crypt.Ciphertext, cek, nonce)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+// TODO: input should be a []byte ( proto, etc )
```
Comment (wibblymat) [inline webpush/encrypt.go:145]: I agree, but it's probably better to raise a GitHub issue rather than add new TODOs.

(I know I already have a couple in the code that I should totally get rid of)

File: webpush/encrypt.go [hunk 3]
```diff
@@ -139,36 +142,64 @@ type EncryptionResult struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	return EncryptWithTempKey(sub, plaintext, serverPrivateKey, serverPublicKey)
+}
+
+// Encrypt a message using Web Push protocol, reusing the temp key.
+// A new salt will be used. This is ~20% faster.
+func EncryptWithTempKey(sub *Subscription, plaintext []byte,
```
Comment (wibblymat) [inline webpush/encrypt.go:167]: Perhaps `EncryptWithKey`?  `...TempKey` makes me think that the key will _not_ be reused.

File: webpush/encrypt.go [hunk 4]
```diff
@@ -139,36 +142,64 @@ type EncryptionResult struct {
...
+	// a fresh private/public key pair at random every time we encrypt.
+	serverPrivateKey, serverPublicKey, err := randomKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return EncryptWithTempKey(sub, plaintext, serverPrivateKey, serverPublicKey)
+}
+
+// Encrypt a message using Web Push protocol, reusing the temp key.
```
Comment (wibblymat) [inline webpush/encrypt.go:165]: Doc comments should start with the exact name of the function. Coupled with the next suggestion:

> EncryptWithKey uses the given key-pair to encrypt a message using the Web Push protocol

File: webpush/encrypt.go (hunk 1 above)
Comment (wibblymat) [inline webpush/encrypt.go]: If the decryption is only being used for testing then move the functions to the tests.

File: webpush/encrypt.go [hunk 5]
```diff
@@ -291,9 +322,30 @@ func encrypt(plaintext, key, nonce []byte) ([]byte, error) {
 		return nil, err
 	}
 
+	// TODO: to reduce allocations, allow out buffer to be passed in
```
Comment (wibblymat) [inline webpush/encrypt.go:312]: Create an issue

File: webpush/encrypt_test.go [hunk 6]
```diff
@@ -151,3 +177,19 @@ func TestEncrypt(t *testing.T) {
 		t.Errorf("Expected salt to be %v, got %v", result.Salt, expSalt)
 	}
 }
+
+func BenchmarkEncrypt(b *testing.B) {
```
Comment (wibblymat) [inline webpush/encrypt_test.go:218]: 👍

File: webpush/push.go [hunk 7]
```diff
@@ -80,6 +116,8 @@ func Send(client *http.Client, sub *Subscription, message, token string) (*http.
 	}
 
 	req, err := NewPushRequest(sub, message, token)
+	// Default TTL
+	req.Header.Add("ttl", "0")
```
Comment (wibblymat) [inline webpush/push.go:124]: Already added in #5

File: webpush/vapid.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"math/big"
+	"net/url"
+	"time"
+)
+
+var (
+	vapidPrefix = []byte("eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.")
```
Comment (wibblymat) [inline webpush/vapid.go:31]: What is this?

File: webpush/vapid.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	dot         = []byte(".")
+)
+
+type jwtPrefix struct {
```
Comment (wibblymat) [inline webpush/vapid.go:37]: Not needed?

File: webpush/vapid.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+		token = append(token, dot...)
+		token = append(token, sigB64...)
+	}
+	res = string(token)
+	return
+}
+
+// NewVapid constructs a new Vapid generator from EC256 public and private keys,
+// in uncompressed format.
+func NewVapid(publicKey, privateKey string) (v *Vapid) {
```
Comment (wibblymat) [inline webpush/vapid.go:115]: I think this should probably accept `[]byte`s rather than `string`s, and not do the base64 decode here.

File: webpush/vapid.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	PublicKey string
+
+	// The private key used to sign tokens
+	pkey *ecdsa.PrivateKey
+
+	// Sub should be an email or URL, for identification
+	Sub string
+}
+
+// TODO: cache for tokens
```
Comment (wibblymat) [inline webpush/vapid.go:62]: Create an issue

File: webpush/vapid.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+// Token creates a token with the specified endpoint, using configured Sub id
+// and a default expiration (1h).
+func (vapid *Vapid) Token(aud string) (res string) {
+	url, _ := url.Parse(aud)
```
Comment (wibblymat) [inline webpush/vapid.go:67]: Don't ignore the error

File: webpush/vapid.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	enc.Encode(t64, t)
+
+	token := make([]byte, len(t)+len(vapidPrefix)+100)
+	token = append(token[:0], vapidPrefix...)
+	token = append(token, t64...)
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(token)
+
+	if r, s, err := ecdsa.Sign(rand.Reader, vapid.pkey, hasher.Sum(nil)); err == nil {
```
Comment (wibblymat) [inline webpush/vapid.go:87]: If I'm reading this correctly, any error here is silently ignored.

File: webpush/realpush_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package webpush
```
Comment (wibblymat) [inline webpush/realpush_test.go:1]: I have a very similar test script on my own machine that I didn't check in, to avoid putting any real subscriptions in the codebase. There is a very minor information leak (especially since mine has my GCM auth key in it) and the details basically only work for a single user, because a full integration test means manually checking that you received a notification.

@gauntface, what is the state of the art for testing push with real browsers? I feel like the node module has something better.

All that said, I guess I'd rather have imperfect integration tests than none.

File: webpush/push.go [hunk 15]
```diff
@@ -69,6 +69,42 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
 	return req, nil
 }
 
+// NewVapidRequest creates a valid Web Push HTTP request for sending a message
+// to a subscriber, using Vapid authentication. You can add more headers to
+// configure collapsing, TTL.
+func NewVapidRequest(sub *Subscription, message string, vapid *Vapid) (*http.Request, error) {
+	// If the endpoint is GCM then we temporarily need to rewrite it, as not all
```
Comment (wibblymat) [inline webpush/push.go:78]: This comment has been copied but doesn't actually apply here

File: webpush/push.go [hunk 16]
```diff
@@ -69,6 +69,42 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
...
+	// future.
+	req, err := http.NewRequest("POST", sub.Endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	tok := vapid.Token(sub.Endpoint)
+	req.Header.Add("Authorization", fmt.Sprintf(`Bearer %s`, tok))
+
+	// If there is no payload then we don't actually need encryption
```
Comment (wibblymat) [inline webpush/push.go:91]: In the VAPID case, I'm not sure this holds true. I think you have to specify the `Crypto-Key: p256ecdsa=...` header even for a tickle if you send the Authorization header. Though reading the spec doesn't really make this 100% clear.

File: webpush/push.go [hunk 17]
```diff
@@ -69,6 +69,42 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
...
+	// If the endpoint is GCM then we temporarily need to rewrite it, as not all
+	// GCM servers support the Web Push protocol. This should go away in the
+	// future.
+	req, err := http.NewRequest("POST", sub.Endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	tok := vapid.Token(sub.Endpoint)
+	req.Header.Add("Authorization", fmt.Sprintf(`Bearer %s`, tok))
```
Comment (wibblymat) [inline webpush/push.go:89]: Spec says `WebPush` rather than `Bearer` - if implementations accept Bearer then that's fine, we just need to be aware it may change.

**Comment by wibblymat** [comment]
@surma, since I'm not a Go master

File: webpush/vapid.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+		// Vapid key is 32 bytes
+		keyBytes := 32
+		sig := make([]byte, 2*keyBytes)
+
+		rBytes := r.Bytes()
```
Comment (wibblymat) [inline webpush/vapid.go:92]: I feel like this could be simpler:

``` go
rBytes := r.Bytes()
copy(sig[keyBytes-len(rBytes):], rBytes)

sBytes := s.Bytes()
copy(sig[2*keyBytes-len(sBytes):], sBytes)

sigB64 := make([]byte, enc.EncodedLen(len(sig)))
enc.Encode(sigB64, sig)
```

Also allocates less space. Prepared to believe I missed something though, as I've tripped over Go's slices before.

File: webpush/push.go [hunk 19]
```diff
@@ -69,6 +69,42 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
 	return req, nil
 }
 
+// NewVapidRequest creates a valid Web Push HTTP request for sending a message
+// to a subscriber, using Vapid authentication. You can add more headers to
+// configure collapsing, TTL.
+func NewVapidRequest(sub *Subscription, message string, vapid *Vapid) (*http.Request, error) {
```
Comment (wibblymat) [inline webpush/push.go]: If possible I'd like NewVapidRequest to be a thin wrapper around NewPushRequest. Is it feasible to call `NewPushRequest` and then manipulate the headers rather than duplicating the code?

File: webpush/encrypt_test.go [hunk 20]
```diff
@@ -151,3 +177,19 @@ func TestEncrypt(t *testing.T) {
...
+	sub, _ := SubscriptionFromJSON(subscriptionJSON)
```
Comment (surma) [inline webpush/encrypt_test.go:219]: Call `b.ResetTimer()` after this to ensure we are only measuring `Encrypt()`.

File: webpush/encrypt_test.go [hunk 21]
```diff
@@ -151,3 +177,19 @@ func TestEncrypt(t *testing.T) {
...
+	for i := 0; i < b.N; i++ {
+		Encrypt(sub, "Hello world")
+	}
+}
+func BenchmarkEncryptWithKey(b *testing.B) {
+	sub, _ := SubscriptionFromJSON(subscriptionJSON)
+	plain := []byte("Hello world")
+	serverPrivateKey, serverPublicKey, _ := randomKey()
```
Comment (surma) [inline webpush/encrypt_test.go:227]: `b.ResetTimer()`

File: webpush/push.go (hunk 17 above)
Comment (surma) [inline webpush/push.go:89]: I’d still vote for sticking to the spec

File: webpush/realpush_test.go (hunk 14 above)
Comment (surma) [inline webpush/realpush_test.go:1]: For tests against live systems, I’d recommend following [12FactorApp](http://12factor.net/) and putting the credentials in environment variables. Skip the affected tests with a warning if these credentials are not present.

File: webpush/vapid.go (hunk 18 above)
Comment (surma) [inline webpush/vapid.go:92]: Don’t even need the assigment to a local variable. 2 `copy()` calls and the last 2 lines. #microoptimizations 😄

**Comment by surma** [comment]
A few remarks from my side, agree with all of @wibblymat’s feedback.

File: webpush/push.go (hunk 17 above)
Comment (wibblymat) [inline webpush/push.go:89]: @surma in this case that's impractical - if the server still expects `Bearer` then sticking to the spec will just mean that it fails :)

File: webpush/encrypt.go (hunk 1 above)
Comment (costinm) [inline webpush/encrypt.go]: This package can be used for sending messages, but it can also be used to implement an UA - the spec doesn't require the UA to be a browser, could be an IoT device.

I'm trying to add the http2 framer and subscribe calls to get to a full-UA status.

File: webpush/encrypt.go (hunk 2 above)
Comment (costinm) [inline webpush/encrypt.go:145]: Would it be ok to remove "TODO" and keep the comment ? I think it's better to use issues for defects or features - small optional internal code improvement ideas would generate too much noise.

File: webpush/encrypt.go (hunk 3 above)
Comment (costinm) [inline webpush/encrypt.go:167]: EphemeralKey ? DHKey ?

File: webpush/push.go (hunk 19 above)
Comment (costinm) [inline webpush/push.go]: I would rather remove NewPushRequest, and rename this method.

NewPushRequest uses a token, which is not in the standard - I can add a 'legacyToken' to vapid class, or as a config somewhere - to allow use with old tokens - but I would rather not have it in the main interface.

File: webpush/push.go (hunk 17 above)
Comment (costinm) [inline webpush/push.go:89]: The spec had Bearer few versions back, I'll check if the new name works with the servers.

File: webpush/push.go (hunk 16 above)
Comment (costinm) [inline webpush/push.go:91]: I'll check, but I think it would be a bug in the server if it requires the ecdsa crypto key without payload.

File: webpush/encrypt.go (hunk 3 above)
Comment (surma) [inline webpush/encrypt.go:167]: Is the encryption process any different depending on the type of key? If not, I’d call it `EncryptWithKey`, or even just `Encrypt`.

File: webpush/encrypt.go (hunk 1 above)
Comment (surma) [inline webpush/encrypt.go]: I _tend_ to agree that this package could use a `Decrypt()` function. However, if there’s any concerns about API contracts or maintainability, I think covering the 99% use-case of sending messages only. 

Your call, @wibblymat.

File: webpush/push.go (hunk 7 above)
Comment (costinm) [inline webpush/push.go:124]: Missed that - it has to be removed ( as well as the todo to make it variable :-)

Since the spec requires it and doesn't have a default, I think it makes sense to have it as a request parameter.

File: webpush/realpush_test.go (hunk 14 above)
Comment (costinm) [inline webpush/realpush_test.go:1]: Sounds good - I'll try to get it working in drone.

File: webpush/encrypt.go (hunk 1 above)
Comment (gauntface) [inline webpush/encrypt.go]: The UA use case feels like a step too far for the core goal of this library. Is there a way to keep this for the sending use case and a seperate module could inherit this module and add on the UA specific bits?

File: webpush/encrypt.go (hunk 4 above)
Comment (gauntface) [inline webpush/encrypt.go:165]: Is there a linter you can use in go to keep this maintained?

File: webpush/vapid.go (hunk 8 above)
Comment (costinm) [inline webpush/vapid.go:31]: Added a comment, filled the jwtPrefix as well.

File: webpush/encrypt.go (hunk 4 above)
Comment (surma) [inline webpush/encrypt.go:165]: `golint` yells at you. Should run `golint` and `govet` and all code :)

File: webpush/realpush_test.go (hunk 14 above)
Comment (gauntface) [inline webpush/realpush_test.go:1]: I started working on selenium tests, but the  libraries are sketchy at best in terms of their implementation and one key feature used in Firefox and Chrome is the ability to set profiles / preferences. There is a Firefox Profile Struct but it has an unusual constructor, asked on the project for clarification, but not hopeful. Will carry on exploring for Go.

File: webpush/vapid.go (hunk 9 above)
Comment (costinm) [inline webpush/vapid.go:37]: I'm planning to also add verification of the vapid - it's work in progress. As I mentioned on decryption comment, the code could be used not only for sending, but also for UA and server implementations.

File: webpush/push.go (hunk 16 above)
Comment (gauntface) [inline webpush/push.go:91]: I imagine in Chrome the crypto-key will still be needed to authenticate the push message is from the application server.

Firefox doesn't currently do anything with the application key but will take both of these headers. If either of them are incorrect it will return an error code.

File: webpush/push.go (hunk 17 above)
Comment (gauntface) [inline webpush/push.go:89]: I've been using Bearer in FF and Chrome and it's been working :-/

Comment (costinm) [inline webpush/push.go:89]: Bearer is working, but the spec changed - I'll check if google server accepts the new one and fix it.

File: webpush/push.go (hunk 16 above)
Comment (costinm) [inline webpush/push.go:91]: Interesting point - encrypting an empty message can still serve for authentication. I don't think current code in chrome does this - the 'empty message' case was added before encryption was implemented, but it should.

File: webpush/vapid.go (hunk 10 above)
Comment (costinm) [inline webpush/vapid.go:115]: I tried that - I could add a method with the []byte, but it seems easier/more natural this way, most of the time the keys are passed around as b64 strings - in headers, etc.

File: webpush/encrypt.go (hunk 4 above)
Comment (costinm) [inline webpush/encrypt.go:165]: I think the drone auto-build is running both - I'll double check.

File: webpush/encrypt.go (hunk 1 above)
Comment (costinm) [inline webpush/encrypt.go]: It would make it a bit tricky as API - I can move the code to a separate file, and keep it only in my fork. Having 2 packages would require exposing too much. I'm trying to make it complete - i.e. including delivery receipts, which involves http2 framer, UA receipt side with push promises, and the basic elements of the server. If this doesn't fit this repo, I can organize the code to keep the extra functionality in separate files in a fork.

Comment (wibblymat) [inline webpush/encrypt.go]: If we have decrypt then it should be in here. Perhaps pull it into a separate PR so that we can discuss it separately?

File: webpush/encrypt.go (hunk 3 above)
Comment (wibblymat) [inline webpush/encrypt.go:167]: Changing the signature of an exported function is bad in Go, since it will break existing consumers, so not `Encrypt`. I definitely think that `EncryptWithKey` is the name that makes sense - Ephemeral is the opposite of the correct meaning, since the whole point is to allow you to _not_ use a key that disappears as soon as you use it.

File: webpush/encrypt.go (hunk 2 above)
Comment (aminitu) [inline webpush/encrypt.go:145]: serverPrivateKey, serverPublicKey, err := randomKey()
-   if err != nil {
-       return nil,

File: webpush/encrypt_test.go (hunk 6 above)
Comment (aminitu) [inline webpush/encrypt_test.go:218]: - #8

**Comment by gauntface** [comment]
Closing this pull request - it's too large and covering too many changes to merge.
